### ISSUE TITLE
Ground truth output loss and dice loss go to zero

### ISSUE THREAD

**Issue by jayurbain** [IAmSuyogJadhav/3d-mri-brain-tumor-segmentation-using-autoencoder-regularization#45]
I would appreciate some feedback on my training curves.  Looks like over-fitting which I thought the VAE would address. L2 and KL are set to 0.1 for the VAE in the loss function. The 'Dec_GT_Output_loss' and 'Dec_GT_Output_dice_coefficient' flat line at zero after ~100 epochs. The VAE portion seems to eventually improve.

After 50 epochs:
![image](https://user-images.githubusercontent.com/3729362/82251358-a7afd180-9912-11ea-9adc-d0a11422afd9.png)

Here's an example ground truth:

![image](https://user-images.githubusercontent.com/3729362/82251542-feb5a680-9912-11ea-9a23-44e12c8c20a6.png)

And the corresponding prediction. Not very good.

![image](https://user-images.githubusercontent.com/3729362/82251598-1856ee00-9913-11ea-8508-a455c0d0abf1.png)

After 100 epochs:

![image](https://user-images.githubusercontent.com/3729362/82311559-817a4800-998b-11ea-84c6-87cac32b5ad7.png)

![image](https://user-images.githubusercontent.com/3729362/82311607-8f2fcd80-998b-11ea-8e00-f6ad9ca2f12f.png)

After 150 epochs:

![image](https://user-images.githubusercontent.com/3729362/82311662-a2429d80-998b-11ea-8a73-b6f7ed2bc801.png)

![image](https://user-images.githubusercontent.com/3729362/82311687-ab336f00-998b-11ea-99df-7283c4951e9b.png)

Thanks,
Jay

**Comment by IAmSuyogJadhav** [comment]
Can you try removing the minus sign from this line and see if it improves the result?
https://github.com/IAmSuyogJadhav/3d-mri-brain-tumor-segmentation-using-autoencoder-regularization/blob/7471e81923741aa1b3593e4e95fc090c2e96c8cc/model.py#L143

**Comment by IAmSuyogJadhav** [comment]
Also, is there a reason why the MRI scans in the second figure are blurry? Are they representing the output of VAE branch?

**Comment by jayurbain** [comment]
Will do.

Thanks,
Jay

**Comment by jayurbain** [comment]
Yes. The second figure are the generated VAE images. I should have labeled
these.

I have another question. Using SimpleITK,  your notebook example reads the
images in as C, D, H, W (4,155,240,240).

The model specifically states that it expects the data as C, H, W, D.
Should this matter? I don't think so as long as you're
consistent.

Thanks,
Jay

**Comment by IAmSuyogJadhav** [comment]
That might be a problem, I created the notebook after I had verified the model to be working on Brats18 dataset, so it might be the case that I accounted for that before but forgot to write the code for it in the notebook.

**Comment by jayurbain** [comment]
I'll move the axis and give it a try.

Thanks,
Jay

**Comment by jayurbain** [comment]
I have additionally tried each of the following:
1) Modified the loss to: return - K.mean(2 * intersection / dn, axis=[0,1]) 
2) Modified the loss to: return 1 - K.mean(2 * intersection / dn, axis=[0,1])  // output shown in image
3) Changed order of dimensions in BraTS from C, D, H, W to C, H, W, D

For these latest experiments I used the following dimensions:
image_.shape (4, 128, 128, 96)
label_.shape (3, 128, 128, 96)

No improvement. The model is just not learning.

![image](https://user-images.githubusercontent.com/3729362/82438899-5bbe7300-9a5f-11ea-8ae3-0748ee71aafb.png)

Any further thoughts would be appreciated. The biggest problem, I believe are with the Dec_GT_Output_dice_coefficient and the Dec_GT_Output_loss. In other words, the basic learning task. What am I missing that you used to run on BraTS? In addition, I ran your example notebook on 50, 100, and 150 and get the same poor results.

 I'm thinking I should just try to learn a 3D Unet without the VAE or at least set its weights very low. Get that to work, then try to add the VAE back in.

Thanks,
Jay
